tools/wptrunner/wptrunner/testloader.py

```diff
@@ -31,29 +32,105 @@ def do_delayed_imports():
     from manifest.download import download_from_github  # type: ignore
 
 
-class TestGroupsFile:
-    """
-    Mapping object representing {group name: [test ids]}
-    """
-
-    def __init__(self, logger, path):
+class TestGroups:
+    def __init__(self, logger, path, subsuites):
         try:
             with open(path) as f:
-                self._data = json.load(f)
+                data = json.load(f)
         except ValueError:
             logger.critical("test groups file %s not valid json" % path)
             raise
 
+        self.tests_by_group = defaultdict(set)
         self.group_by_test = {}
-        for group, test_ids in self._data.items():
+        for group, test_ids in data.items():
             for test_id in test_ids:
-                self.group_by_test[test_id] = group
+                id_parts = group.split(":", 1)
+                if len(id_parts) == 1:
+                    group_name = id_parts[0]
+                    subsuite = None
+                else:
+                    subsuites, group_name = id_parts
+                elif subsuite not in subsuites:
+                    raise ValueError(f"Unknown subsuite {subsuite} in group data {group}")
+                self.group_by_test[(subsuite, test_id)] = group_name
+                self.tests_by_group[group_name].add(test_id)
+
+
+def load_subsuites(logger, base_run_info, path: str | None, include_subsuites: set[str]):
+    subsuites: Mapping[str, Subsuite] = {}
+    run_all_subsuites = not include_subsuites
+    now = datetime.now()
+
+    def maybe_add_subsuite(name, data):
+        if run_all_subsuites or name in include_subsuites:
+            if data.get("expires"):
+                if datetime.fromisoformat(data["expires"]) < now:
+                    logger.warning(f"Tried to run expired subsuite {name}")
+                    return
+            subsuites[name] = Subsuite(name,
+                                       base_run_info,
+                                       config=data.get("config", {}),
+                                       run_info=data.get("run_info", {}),
+                                       include=data.get("include"),
+                                       tags=set(data["tags"]) if "tags" in subsuite else None)
+            if name in include_subsuites:
+                include_subsuites.remove(name)
+
+    maybe_add_subsuite("", {})
```

Comment:
"" is not in `include_subsuites`, so this is essentially a no-op if `sun_all_subsuite` is not True. I think we need to append "" to `include_subsuites`?

Reply:
Looks like this is now obsolete.